File: problems/1001-1100/1034_outline-a-color-region/solution.py
from typing import List, Tuple


class Solution:
    def outlineRegion(self, grid: List[List[int]], row: int, col: int, color: int) -> List[List[int]]:
        # Identify the whole connected component first (BFS with an
        # explicit queue — depth safety), classifying each member's
        # border status against the ORIGINAL grid values. Only after
        # every member has been classified does a second pass repaint
        # the collected border cells, so an in-progress repaint can
        # never corrupt a later cell's neighbor check.
        m, n = len(grid), len(grid[0])
        original = grid[row][col]
        visited = [[False] * n for _ in range(m)]
        visited[row][col] = True
        queue: List[Tuple[int, int]] = [(row, col)]
        head = 0
        border: List[Tuple[int, int]] = []
        while head < len(queue):
            r, c = queue[head]
            head += 1
            is_border = r == 0 or r == m - 1 or c == 0 or c == n - 1
            for nr, nc in ((r - 1, c), (r + 1, c), (r, c - 1), (r, c + 1)):
                if 0 <= nr < m and 0 <= nc < n:
                    if grid[nr][nc] != original:
                        is_border = True
                    elif not visited[nr][nc]:
                        visited[nr][nc] = True
                        queue.append((nr, nc))
            if is_border:
                border.append((r, c))
        for r, c in border:
            grid[r][c] = color
        return grid

File: problems/1001-1100/1034_outline-a-color-region/solution.ts
// Identify the whole connected component first (BFS with an explicit
// queue — depth safety), classifying each member's border status against
// the ORIGINAL grid values. Only after every member has been classified
// does a second pass repaint the collected border cells, so an
// in-progress repaint can never corrupt a later cell's neighbor check.
function outlineRegion(grid: number[][], row: number, col: number, color: number): number[][] {
    const m = grid.length;
    const n = grid[0].length;
    const original = grid[row][col];
    const visited: boolean[] = new Array(m * n).fill(false);
    const queue: number[] = new Array(m * n).fill(0);
    const border: number[] = new Array(m * n).fill(0);
    let borderCount = 0;
    let head = 0;
    let tail = 0;
    visited[row * n + col] = true;
    queue[tail] = row * n + col;
    tail += 1;
    while (head < tail) {
        const cell = queue[head];
        head += 1;
        const r = Math.floor(cell / n);
        const c = cell % n;
        let isBorder = r === 0 || r === m - 1 || c === 0 || c === n - 1;
        if (r > 0) {
            if (grid[r - 1][c] !== original) {
                isBorder = true;
            } else if (!visited[cell - n]) {
                visited[cell - n] = true;
                queue[tail] = cell - n;
                tail += 1;
            }
        }
        if (r + 1 < m) {
            if (grid[r + 1][c] !== original) {
                isBorder = true;
            } else if (!visited[cell + n]) {
                visited[cell + n] = true;
                queue[tail] = cell + n;
                tail += 1;
            }
        }
        if (c > 0) {
            if (grid[r][c - 1] !== original) {
                isBorder = true;
            } else if (!visited[cell - 1]) {
                visited[cell - 1] = true;
                queue[tail] = cell - 1;
                tail += 1;
            }
        }
        if (c + 1 < n) {
            if (grid[r][c + 1] !== original) {
                isBorder = true;
            } else if (!visited[cell + 1]) {
                visited[cell + 1] = true;
                queue[tail] = cell + 1;
                tail += 1;
            }
        }
        if (isBorder) {
            border[borderCount] = cell;
            borderCount += 1;
        }
    }
    for (let i = 0; i < borderCount; i++) {
        const cell = border[i];
        grid[Math.floor(cell / n)][cell % n] = color;
    }
    return grid;
}
